appserver/ejb/ejb-rest-remoting/client/src/main/java/fish/payara/ejb/rest/client/RemoteEJBContext.java

```diff
@@ -0,0 +1,382 @@
+ * Times are in microseconds, and values can either be given as Strings, number values or object instances.
+ * 
+ * @author Arjan Tijms
+ * @since Payara 5.191
+ *
+ */
+class RemoteEJBContext implements Context {
+    
+    private Hashtable<String, Object> environment;
+    
+    @SuppressWarnings("unchecked")
+    public RemoteEJBContext(Hashtable<?, ?> environment) {
+        this.environment = (Hashtable<String, Object>) environment;
+    }
+
+    @Override
+    public Object lookup(Name name) throws NamingException {
+        return lookup(name.toString());
+    }
+
+    @Override
+    public Object lookup(String name) throws NamingException {
+        String url = (String) environment.get(PROVIDER_URL);
+        
+        try {
+            // Get client build with all optional config applied
+            ClientBuilder clientBuilder = getClientBuilder();
+            
+            // For the lookup do a call to the remote server first to obtain
+            // the remote business interface class name given a JNDI lookup name.
+            // The JNDI lookup name normally does not give us this interface name.
+            // This also allows us to check the JNDI name indeed resolves before
+            // we create a proxy and return it here.
+            
+            Map<String, Object> payload = new HashMap<>();
+            payload.put("lookup", name);
+            
+            URI remotePayaraURI = new URL(url).toURI();
+            
+            String className = 
+                clientBuilder
+                    .build()
+                    .target(remotePayaraURI)
+                    .path("ejb")
+                    .path("lookup")
+                    .request()
+                    .post(Entity.entity(payload, APPLICATION_JSON))
+                    .readEntity(String.class);
+            
+            // After we have obtained the class name of the remote interface, generate
+            // a proxy based on it.
+            
+            return EjbRestProxyFactory.newProxy(
+                    Class.forName(className), 
+                    clientBuilder.build()
+                                 .target(remotePayaraURI)
+                                 .path("ejb")
+                                 .path("invoke"), 
+                    name,
+                    new HashMap<String, Object>(environment)
+                    );
+            
+        } catch (Exception e) {
+            throw newNamingException(name, e);
+        }
+    }
+    
+    private ClientBuilder getClientBuilder() throws ClassNotFoundException, InstantiationException, IllegalAccessException {
+        ClientBuilder clientBuilder = ClientBuilder.newBuilder();
+        
+        if (environment.containsKey("fish.payara.connectTimeout")) {
+            clientBuilder.connectTimeout(getLong(environment.get("fish.payara.connectTimeout")).longValue(), MICROSECONDS);
```

Comment:
These would probably be better as final static variables or enums

Reply:
Agreed